### PR TITLE
Add RecyclerView's ListAdapter extension for renderers

### PR DESCRIPTION
**Pull request by vjgarciag96**
I'm not sure if this library still accepts pull requests, but I'll try. 

In this PR you'll see the changes needed to add a new extension to provide a renderers friendly adapter for the new recycler view `ListAdapter` (https://developer.android.com/reference/android/support/v7/recyclerview/extensions/ListAdapter). This `ListAdapter` is a convenience wrapper around `AsyncListDiffer` that implements Adapter common default behavior. `AsyncListDiffer` is a helper for computing the difference between two lists via `DiffUtil` in a background thread. 

So, the main advantage of this adapter vs `RecyclerView.Adapter` is the diffing computation on a background thread. I think it could be useful to provide it from the library.

Let's go with the changes:

- In `RVListRendererAdapter` file you'll see the code of the new adapter. It's basically the same code of `RVRendererAdapter` with some changes in constructors and diffing. 
`ListAdapter` needs to receive the diffing configuration when it's instantiated and that configuration can be provided as a `DiffUtil.ItemCallback` implementation for the adapter items type or as `AsyncListDiffer` which is a more advanced configuration where you can even configure the thread executor where the diffing is going to be executed. We have created the constructors needed to support this.
To execute the diffing on the adapter we have added the `submitList` method. This method overrides `ListAdapter`'s own  implementation and create a renderers friendly implementation for it (it's needed to override this method to avoid the possibility of accessing its `ListAdapter` implementation from outside through `RVListRendererAdapter`).
- Regarding tests, a new test suite has been added for `RVListRendererAdapter`. It's basically the same as the `RVRendererAdapter` one adapted to the new class.
- Moreover, a sample has been added for the new adapter. The sample is really simple, we set up the recycler view and the adapter and we populate it with a random list of videos. Perhaps I should provide a more detailed sample with diffing calculation. 
- I haven't added documentation because I don't see any documentation related with the different adapter implementations provided by the library, but if you think it's needed, I could add it without problems.

There's a lot of code duplication on these new changes. I thought about extracting common code in some way to reuse it, but it wasn't easy and I'm not sure if it worths it. We can discuss about about it if you want.

Any comments, questions or suggestions are welcome. Thanks!

### CODE REVIEW

File: renderers/src/main/java/com/pedrogomez/renderers/RVListRendererAdapter.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+ * information into a RecyclerView. The main difference between RVRendererAdapter and RVListRendererAdapter
+ * is that RVListRendererAdapter performs the adapter elements' diffing calculation in a background thread.
+ * <p>
+ * RVListRendererAdapter has to be constructed with a RendererBuilder to provide Renderer
+ * to RvListRendererAdapter, one of DiffUtil.ItemCallback or AsyncDifferConfig to provide the
+ * diffing configuration and one AdapteeCollection to provide the elements to render.
+ *
+ * @author Víctor Julián García Granado.
+ */
+public class RVListRendererAdapter<T> extends ListAdapter<T, RendererViewHolder> {
```
Comment (vjgarciag96) [inline renderers/src/main/java/com/pedrogomez/renderers/RVListRendererAdapter.java]: It's basically the same code as `RVRendererAdapter` with changes on constructors and diffing calculation.

File: renderers/src/test/java/com/pedrogomez/renderers/RVListRendererAdapterTest.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import org.robolectric.annotation.Config;
+
+import java.util.Collection;
+import java.util.LinkedList;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.notNull;
+import static org.mockito.Mockito.*;
+
+@Config(sdk = 16) @RunWith(RobolectricTestRunner.class) public class RVListRendererAdapterTest {
```
Comment (vjgarciag96) [inline renderers/src/test/java/com/pedrogomez/renderers/RVListRendererAdapterTest.java]: Same tests as `RvRendererAdapter` ones adapter to the new adapter

File: sample/src/main/java/com/pedrogomez/renderers/sample/ui/diffing/VideoItemDiffCallback.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.pedrogomez.renderers.sample.ui.diffing;
+
+import android.support.annotation.NonNull;
+import android.support.v7.util.DiffUtil;
+import com.pedrogomez.renderers.sample.model.Video;
+
+public class VideoItemDiffCallback extends DiffUtil.ItemCallback<Video> {
```
Comment (pedrovgs) [inline sample/src/main/java/com/pedrogomez/renderers/sample/ui/diffing/VideoItemDiffCallback.java]: Couldn't we provide a default implementation based on the equality of the instances so the user doesn't need to implement this ``DiffCallback`` by default? I think we already did this with the ``diffUpdate`` implementation the ``RVRendererAdaper`` has. The user should be able to override this behavior, but by default we could provide this implementation based on a generic T instead.

Comment (pedrovgs) [reply sample/src/main/java/com/pedrogomez/renderers/sample/ui/diffing/VideoItemDiffCallback.java]: If we are able to add this default implementation as the last parameter we could easily replace the RVRendererAdapter with RVListRendererAdapter by just changing the constructor.

Comment (vjgarciag96) [reply sample/src/main/java/com/pedrogomez/renderers/sample/ui/diffing/VideoItemDiffCallback.java]: Yes, we can! Then, do you think that the new adapter implementation should replace the older one (as it's going to keep the same signature on constructors) or that both implementations should be kept? 
I have added the changes with the default items' diffing implementation keeping both adapters, but we can change if you consider we should only have one.

File: renderers/src/main/java/com/pedrogomez/renderers/DefaultDiffUtilItemCallback.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.pedrogomez.renderers;
+
+import android.support.annotation.NonNull;
+import android.support.v7.util.DiffUtil;
+
+public class DefaultDiffUtilItemCallback<T> extends DiffUtil.ItemCallback<T> {
```
Comment (vjgarciag96) [inline renderers/src/main/java/com/pedrogomez/renderers/DefaultDiffUtilItemCallback.java]: The new default implementation for the items' diffing

File: renderers/src/main/java/com/pedrogomez/renderers/RVListRendererAdapter.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+    public RVListRendererAdapter(RendererBuilder<T> rendererBuilder) {
+        this(rendererBuilder, new DefaultDiffUtilItemCallback<T>(), new ListAdapteeCollection<T>());
+    }
+
+    public RVListRendererAdapter(RendererBuilder<T> rendererBuilder, AdapteeCollection<T> collection) {
+        super(new DefaultDiffUtilItemCallback<T>());
+        this.rendererBuilder = rendererBuilder;
+        this.collection = collection;
+    }
```
Comment (vjgarciag96) [inline renderers/src/main/java/com/pedrogomez/renderers/RVListRendererAdapter.java]: The new constructors with the default diffing implementation

File: README.md [hunk 6]
```diff
@@ -110,6 +110,10 @@ adapter.diffUpdate(newList)
...
 
+Also, `RVListRendererAdapter` provides a way to perform diff updates in a background thread transparently. When using `RVListRendererAdapter` you'll have a default `DiffUtil.ItemCallback` implementation (https://developer.android.com/reference/android/support/v7/util/DiffUtil.ItemCallback)) based on referencial equality for `areItemsTheSame` method and structural equality for `areContentsTheSame` method. You also have constructors on this class to provide your own implementation for `DiffUtil.ItemCallback`. You can even configure the threads used to perform the calculations through `AsynDifferConfig` class (https://developer.android.com/reference/android/support/v7/recyclerview/extensions/AsyncDifferConfig).
+
```
Comment (vjgarciag96) [inline README.md]: New documentation in the README about the new way provided to perform diff updates.

File: sample/src/main/java/com/pedrogomez/renderers/sample/ui/diffing/VideoItemDiffCallback.java (hunk 3 above)
Comment (pedrovgs) [reply sample/src/main/java/com/pedrogomez/renderers/sample/ui/diffing/VideoItemDiffCallback.java]: 🤔I think we can keep it as is right now for the next minor release and I'll review if we can merge them soon.

File: renderers/src/main/java/com/pedrogomez/renderers/DefaultDiffUtilItemCallback.java (hunk 4 above)
Comment (pedrovgs) [reply renderers/src/main/java/com/pedrogomez/renderers/DefaultDiffUtilItemCallback.java]: Great thanks!

File: renderers/src/main/java/com/pedrogomez/renderers/RVListRendererAdapter.java (hunk 5 above)
Comment (pedrovgs) [reply renderers/src/main/java/com/pedrogomez/renderers/RVListRendererAdapter.java]: Thanks!

File: README.md (hunk 6 above)
Comment (pedrovgs) [reply README.md]: Great!

**Pull request merged**